accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+
+    // For ease-of-use and performance, use GetValue().
+    // It may return the raw internal value or the last converted value instead of deserializing.
+    str = node.GetValue<string>();
+```
+
+### Cast operators
+No explicit or implicit cast operators exist since `GetValue<T>` allows for a single, consistent programming model. This also works on not just the "known" types in the CLR but also custom data types known by the serializer.
+
+Cast operators do not work in all languages such as F# and even for C# would not work nicely since `System.Object` is allowed for `JsonArray` and `JsonCollection` elements. For example:
+```cs
+JsonArray jArray = ...
+jArray[2] = "hello";
+// Is jArray[2] a string or a JsonValue?
+
+jArray[2] = 2;
+// jArray[2] is always an int since we can't have implicit cast operators for number types.
+```
+
+Although it is possible to add cast operators for `JsonValue` to\from string\bool, number types should not have any implicit operator since they may throw - for example, an internal value of "3.14" cannot be converted into `int` so an exception must be thrown.
+
+### Specifying the `JsonSerializerOptions` instance
+The `JsonSerializerOptions` instance needs to be specified in order for any run-time added custom converters to be found and to use `PropertyNameCaseInsensitive` for property lookup with `JsonObject`.
+
+It is cumbersome and error-prone to specify the options instance for every new element or property:
+```cs
+    var jObject = new JsonObject(options)
+    {
+        ["MyString"] = new JsonValue("Hello!", options),
+        ["MyBoolean"] = new JsonValue(false, options),
+        ["MyArray"] = new JsonArray(options)
+        {
+            new JsonValue(2, options),
+            new JsonValue(3, options),
+            new JsonValue(42, options)
+        },
+    }
+```
+so it is possible to omit the options instance for non-root members. When a property or element is added, the options instance is set to the parent's instance:
+```cs
+    var jObject = new JsonObject(options) // options still needed at root level
+    {
+        ["MyString"] = new JsonValue("Hello!"),
+        ["MyBoolean"] = new JsonValue(false),
+        ["MyArray"] = new JsonArray()
+        {
+            new JsonValue(2),
+            new JsonValue(3),
+            new JsonValue(42)
+        },
+    }
+```
+
+Pending design, it may be required that all option instances are the same across a given `JsonNode` tree.
+
+### Using `System.Object` for values
+Both `JsonObject` and `JsonArray` allow `System.Object` to be specified for the property\array values. This allows a terse and more performant model, plus compatibility with `dynamic`:
+```cs
+    var jObject = new JsonObject(options)
+    {
+        ["MyString"] = "Hello!",
+        ["MyBoolean"] = false,
+        ["MyArray"] = new int[] {2, 3, 42};
+    }
+```
+
+The values remain in the original form and are not converted to `JsonNode` instances:
```

Comment:
To me, this feels unintuitive. I think these types should have consistent behavior when parsing JSON or creating a tree dynamically.

Reply:
You can still use 100% `JsonNode` types in "edit mode". However, the design doesn't force that because:
- We use `object` instead of `JsonNode` in `IDictionary`\`IList` for less issues with `dynamic` per https://github.com/dotnet/designs/pull/163/files#diff-34fc37119c1d4491b844207aeed3a9c846c3e2d835fd712e164b1e47b89687c0R358
- Consistent with Newtonsoft
- More intuitive in some cases since a literal `true` or `false` can be used (without implicit operators) instead of `JsonBoolean(true)` and `JsonBoolean(false)`.
- More performant that being forced to use `JsonNode` types (removes a somewhat unnecessary abstraction)

The disadvantage of the current design then is that the object model could be a mix of different types, which may be confusing to any code that needed to walk that object model in "edit mode". Note that after deserialization all types are `JsonNode`-derived. However, if such code that needed to walk an object model also needed to operate on CLR types in general, then this wouldn't be an issue.

Reply:
As noted elsewhere, `JsonNode` is now required (not `System.Object`) for array elements and object properties. The raw (or current) value for a node is exposed as `JsonNode.Value` which, for example, would be a bool, string, collection, etc. 

The Value property is mutable (has a setter) and can be set to "anything". When `JsonValue` is serialized that raw value is serialized using the existing serialization semantics (using the options to control formatting, naming policies, custom converters, etc).